lib/beacon_web/controllers/api/component_controller.ex

```diff
@@ -0,0 +1,67 @@
+defmodule BeaconWeb.API.ComponentController do
+  use BeaconWeb, :controller
+  require Logger
+  alias Beacon.ComponentCategories
+  alias Beacon.ComponentDefinitions
+  alias Beacon.ComponentInstances
+  alias Ecto.UUID
+
+  @tag_for_name %{
+    "title" => "h1",
+    "paragraph" => "p",
+    "link" => "a",
+    "button" => "button",
+    "aside" => "aside"
+  }
+  action_fallback BeaconWeb.API.FallbackController
+
+  def index(conn, _params) do
+    component_categories = ComponentCategories.list_component_categories()
+    component_definitions = ComponentDefinitions.list_component_definitions()
+
+    render(conn, :index,
+      component_categories: component_categories,
+      component_definitions: component_definitions
+    )
+  end
+
+  @spec create(Plug.Conn.t(), map) :: Plug.Conn.t()
+  def create(conn, %{"definitionId" => component_definition_id, "pageId" => page_id, "attributes" => attributes}) do
+    definition = ComponentDefinitions.get_component_definition!(component_definition_id)
+    component_data = build_component(definition.blueprint)
+    component_instance = ComponentInstances.create_component_instance!(%{data: component_data, page_id: page_id})
+    render(conn, :show, component: component_instance)
+  end
+
+  def create(conn, %{"definitionId" => component_definition_id, "attributes" => attributes}) do
+    definition = ComponentDefinitions.get_component_definition!(component_definition_id)
+    component_data = build_component(definition.blueprint)
+    render(conn, :show, component: %{id: UUID.generate(), data: component_data})
+  end
+
+  @spec update(Plug.Conn.t(), map) :: Plug.Conn.t()
+  def update(conn, %{"id" => id} = map) do
+    data = Map.delete(map, "id")
+
+    case ComponentInstances.get_component_instance!(id) |> ComponentInstances.update_component_instance_data(data) do
```

Comment:
```suggestion
    instance = ComponentInstances.get_component_instance!(id)
    
    case ComponentInstances.update_component_instance_data(instance, data) do
```
Generally pipes should not be used inside a `case` head.  Instead, do all the work beforehand except for the last line, assign that result to a variable, and then put the last line inside the `case` head